redo `sinpi` and `cospi` kernel polynomial approximation

nsajko opened the pull request:
* Make the `Float32` variants use the same evaluation scheme already used for `Float64`, instead of computing in a wider arithmetic and then rounding back to the narrower arithmetic. This of course results in some loss of accuracy, however it seems more appropriate not to rely on other arithmetics, I guess. Even though the accuracy is worse, the result still seems experimentally to be faithful for each argument from the domain of the kernel, with the ULP error is below `1`.

* Regenerate the polynomials for `Float64`, too, improving accuracy. The degrees of the polynomials stay the same.

Max errors over the domain (`0` to `0.25`) of the kernels:

```none
before
  func = cospi
    type = Float32
      (a, b) = (0.0, 0.25)
      max = 0.50000286f0
    type = Float64
      (a, b) = (0.0, 0.25)
      max = 0.90289307f0
  func = sinpi
    type = Float32
      (a, b) = (0.0, 0.25)
      max = 0.5000706f0
    type = Float64
      (a, b) = (0.0, 0.25)
      max = 0.6906576f0

after
  func = cospi
    type = Float32
      (a, b) = (0.0, 0.25)
      max = 0.8907261f0
    type = Float64
      (a, b) = (0.0, 0.25)
      max = 0.9042959f0
  func = sinpi
    type = Float32
      (a, b) = (0.0, 0.25)
      max = 0.9644079f0
    type = Float64
      (a, b) = (0.0, 0.25)
      max = 0.6577196f0
```

nsajko committed:
redo `sinpi` and `cospi` kernel polynomial approximation

* Make the `Float16` and `Float32` variants use the same evaluation
  scheme already used for `Float64`, instead of computing in a wider
  arithmetic and then rounding back to the narrower arithmetic. This of
  course results in some loss of accuracy, however it seems more
  appropriate not to rely on other arithmetics, I guess. Even though
  the accuracy is worse, the result still seems experimentally to be
  faithful for each argument from the domain of the kernel in each case
  except for `sinpi(::Float16)` which has max error of about 1.4 ULP.

* Regenerate the polynomials for `Float64`, too, improving accuracy.
  The degrees of the polynomials stay the same.

Closes #58977

nsajko commented:
My laptop is currently running a search for max ULP errors which will last several hours.

oscardssmith commented:
Also, IMO doing doublefloat Float16 almost certainly isn't worth it. No hardware that I'm aware of has slow enough Float32 performance to make that worthwhile.

vchuravy commented:
> Also, IMO doing doublefloat Float16 almost certainly isn't worth it. No hardware that I'm aware of has slow enough Float32 performance to make that worthwhile.

:trollface: future GPU/TPUs

oscardssmith commented:
@vchuravy even there, are there systems where FP16 throughput is >4x FP32?

nsajko committed:
fix multiline comment syntax

nsajko committed:
revert changes related to `Float16`, as per oscardssmith

A future PR or a future commit in this PR could also regenerate the
polynomials for the wide kernels for `Float16` and `Float32`.

nsajko committed:
increase by one the degree of the polynomial for `sinpi`, `Float32`

Get the ULP error to below one.

nsajko commented:
The accuracy for `cospi(::Float64)` seems like it may actually have regressed due to the coefficients being regenerated with Sollya, from `0.90289307f0` ULP to `0.9042959f0` ULP. That said, I suppose it's just a measurement error due to not being able to search through the `Float64` space exhaustively.

nsajko commented:
A single test fails:

```none
Error in testset math:
Test Failed at /cache/build/tester-amdci5-9/julialang/julia-master/julia-c968bd880c/share/julia/test/math.jl:634
  Expression: ≈(cosc(0.14), -0.4517331883801308, rtol = 1.0e-15)
   Evaluated: -0.4517331883801315 ≈ -0.4517331883801308 (rtol=1.0e-15)
ERROR: LoadError: Test run finished with errors
in expression starting at /cache/build/tester-amdci5-9/julialang/julia-master/julia-c968bd880c/share/julia/test/runtests.jl:99
```

It tests the accuracy of `cosc(0.14::Float64)`. The relevant `cosc` method is currently implemented in terms of `cospi` and `sinpi`:

https://github.com/JuliaLang/julia/blob/c1c091e45bba36272fac308c7fdfa6c803a22ce5/base/special/trig.jl#L1129-L1136

At `x = 0.14` this PR indeed returns a slightly less accurate result than `master` for `sinpi`: the error goes from `0.49043456f0` ULP to `0.5095654f0` ULP.

It could be argued the test is overly strict, however I'll go and reimplement `cosc` for `Float32` and `Float64` instead of relaxing it.

giordano commented:
> even there, are there systems where FP16 throughput is >4x FP32?

I believe that's the case already for TPUs?

nsajko commented:
Regarding `Float16`, I basically gave up on that because getting the error to below 1 ULP without reaching out to a wider type seems challenging.

nsajko committed:
update doc test, the value is now more accurate

The old `sinpi` value had an error of 0.56462383 ULP, now it's just
0.4353762 ULP. Given that the result is more accurate now, I just
updated the doc test.

nsajko committed:
relax `cispi` doc test

Relax the doc test, don't check exact floating-point equality.

nsajko committed:
just forgo the doctest

nsajko committed:
add some accuracy tests for the kernel functions

nsajko committed:
Merge branch 'master' into sinpi_cospi_kernel_polynomial_approximations

nsajko committed:
address reviewer comment: avoid new structs

nsajko committed:
try to fix test suite on linux32 (no platform FMA support)

nsajko committed:
Merge branch 'master' into sinpi_cospi_kernel_polynomial_approximations

nsajko committed:
Merge branch 'master' into sinpi_cospi_kernel_polynomial_approximations
